### PR TITLE
gh-127586: properly restore blocked signals in resource_tracker.py

### LINKED ISSUE #127586: `multiprocessing.Pool` does not properly restore blocked signals
# Bug report

### Bug description:

When instantiating a `multiprocessing.Pool`, for start methods "spawn" and "forkserver", the resource tracker must register a signal mask in the parent process to block some signals prior to creating the spawned childs, as explained by the comment below:

https://github.com/python/cpython/blob/bfb0788bfcaab7474c1be0605552744e15082ee9/Lib/multiprocessing/resource_tracker.py#L152-L164

The comment makes sense, however the way in which the mask is unregistered seems like a bug, because it can have unintended effects on the parent process.

```python
import signal
from multiprocessing import Pool, set_start_method

def get_blocked_signals():
    return signal.pthread_sigmask(signal.SIG_BLOCK, {})

if __name__ == "__main__":
    set_start_method("spawn")
    signal.pthread_sigmask(signal.SIG_BLOCK, {signal.SIGTERM})
    print(f"I am blocking {get_blocked_signals()}")
    my_pool = Pool(5)
    print(f"After creating Pool, now I am blocking {get_blocked_signals()}")
```

Running this code on the latest commit here, the output is as follows:

```
I am blocking {<Signals.SIGTERM: 15>}
After creating Pool, now I am blocking set()
```

This does not match what I would expect to see, I would have expected the set of blocked signals to remain unchanged after the call to construct the Pool. As far as I know I don't think Pool is supposed to have side effects on the signal handling of the main process.

It appears that in the resource tracker, after spawning the child process, the tracker unblocks the ignored signals rather than restore to the previous sigmask. This works fine for most use cases but has the unintended side effect of unblocking any of the ignored signals that were already blocked prior (as shown in the example above).

This seems relatively trivial to fix so I should have a PR up shortly.

One important thing to note here is with the fix, child processes created by spawn or forkserver will now inherit blocked signals such as SIGTERM, if they are blocked in the parent process. This is already the preexisting behavior for signals that weren't ignored by the resource manager, such as SIGUSR1. Calling methods like terminate(), or garbage collecting the Pool, will hang due to the child processes blocking SIGTERM. In this case Pool cleanup should be handled in the parent using close() followed by join(), or when the Pool is created an initializer is provided in the constructor to unblock SIGTERM in the child. To me this feels more in line with expected behavior.

It might be worthwhile to either adjust Pool.terminate() to call kill() (i.e. SIGKILL) instead on the workers or, to avoid breaking compatibility, provide an implementation of kill() in Pool to match parity with Process, but I'll leave that up for discussion for now.

Please let me know if anything here seems inaccurate or if there are any questions.

### CPython versions tested on:

3.9, 3.10, 3.11, 3.12, 3.13, CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-127587
* gh-127973
* gh-127983
* gh-128011
* gh-128298
* gh-128299

### PR DESCRIPTION
**Pull request by stephen-hansen**
Closes #127586. Instead of using SIG_UNBLOCK to undo the blocked "ignored signals" by resource_manager when creating a child process, we instead will save the blocked signal mask as it was prior to the call to SIG_BLOCK and restore from there using SIG_SETMASK. This has the intended effect of preserving the SIG_BLOCK on either SIGTERM or SIGINT if either was already blocked when entering this method in the parent thread. Existing behavior where either signal was not blocked prior to entering the resource tracker is still preserved since the SIG_SETMASK will undo the SIG_BLOCKs anyway if those signals were previously unblocked.

Please let me know if anything else is needed here, e.g. if a unit test is worthwhile here, I could take a stab tomorrow at writing one.

* Issue: gh-127586

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-12-03-20-28-08.gh-issue-127586.zgotYF.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+:class:`multiprocessing.pool.Pool` now properly restores blocked signal handlers
+of the parent thread when creating processes via either *spawn* or
+*forkserver*. Previously it would unblock **SIGTERM** and **SIGINT** in the
+parent thread on construction, even if either signal handler was
+intentionally blocked prior to the call to construct. Additionally, child
+processes can now inherit said blocked signals.
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-12-03-20-28-08.gh-issue-127586.zgotYF.rst]: Love the effort, but blurbs don't need to be nearly this long. Just the first sentence is fine :)
```suggestion
:class:`multiprocessing.pool.Pool` now properly restores blocked signal handlers
of the parent thread when creating processes via either *spawn* or
*forkserver*.
```

**Comment by ZeroIntensity** [comment]
> Please let me know if anything else is needed here, e.g. if a unit test is worthwhile here, I could take a stab tomorrow at writing one.

That would be great. We generally require tests for all bugfixes.

**Comment by stephen-hansen** [comment]
> > Please let me know if anything else is needed here, e.g. if a unit test is worthwhile here, I could take a stab tomorrow at writing one.
> 
> That would be great. We generally require tests for all bugfixes.

perfect, thanks. I've added a blocked signals test for resource tracker which should cover the issue.

One weird thing I've noticed is that on some of my build pipelines, I'm seeing random test failures for idlelib on macos-13, e.g. https://github.com/python/cpython/actions/runs/12301168597/job/34331141506?pr=127587. Some of the test failures appear similar to what was reported on https://github.com/python/cpython/issues/76152. Not sure if there is an implicit dependency between resource_tracker and those unit tests, that I broke, or if this is unrelated to my changes. I looked for a while but couldn't find any explanation. I also just don't know idlelib too well.

**Comment by ZeroIntensity** [comment]
Yeah, those don't look related. Sometimes we get intermittent failures on some tests, generally not because of a bug in the code but because of a missing platform constraint on that test. I'm not sure if you have permissions to rerun failed CI jobs, so tag me if it happens and I'll do it.

File: Lib/test/_test_multiprocessing.py [hunk 2]
```diff
@@ -6045,6 +6045,21 @@ def test_resource_tracker_exit_code(self):
                     cleanup=cleanup,
                 )
```
Comment (ZeroIntensity) [inline Lib/test/_test_multiprocessing.py]: I suspect we need something like this for some of the buildbots:
```suggestion
    @unittest.skipUnless(hasattr(signal, pthread_sigmask), "pthread_sigmask is not available")
```

**Review by gpshead — APPROVED** [review]

File: Lib/test/_test_multiprocessing.py [hunk 3]
```diff
@@ -6044,6 +6044,21 @@ def test_resource_tracker_exit_code(self):
                 self._test_resource_tracker_leak_resources(
                     cleanup=cleanup,
                 )
+    @unittest.skipUnless(hasattr(signal, pthread_sigmask), "pthread_sigmask is not available")
```
Comment (gpshead) [inline Lib/test/_test_multiprocessing.py]:
```suggestion
    @unittest.skipUnless(hasattr(signal, "pthread_sigmask"), "pthread_sigmask is not available")
```

**Comment by hugovk** [comment]
The next 3.14 alpha release is tomorrow and these tier 1 and 2 buildbot failures are blocking it:

https://buildbot.python.org/#/release_status

Do you have an idea how to fix it or shall we revert this for now?

**Comment by hugovk** [comment]
Revert PR in case we need it: https://github.com/python/cpython/pull/127983

**Comment by ZeroIntensity** [comment]
I swear that I'm a buildbot failure magnet :smile:

**Comment by stephen-hansen** [comment]
Yeah, it looks like the cleanup test behavior might have changed. At first glance it looks like the previous behavior was that both semaphores would leak, guessing terminate() + wait() was immediate? Feels like it is now blocking on the subprocess, as name1 leaks as expected but name2 appears to be cleaned by its finalizer.. this feels correct to me but would appreciate if someone could confirm.

Guessing the fix is to change the 1 expected output line but for now happy to revert this. I’m going to investigate this more tonight.

**Comment by gpshead** [comment]
thanks!

**Pull request merged**